src/sources/index.test.ts

```diff
@@ -0,0 +1,47 @@
+import { getComponents } from './index'
+
+describe('Sources', () => {
+  describe('getComponents', () => {
+    it('handles errors', async () => {
+      const sources = {
+        success1: () => 'foo',
+        throwsErrorObject: () => { throw new Error('bar') },
+        throwsErrorString: () => { throw 'baz' },
+        success2: () => 'baq',
+      }
+      const components = await getComponents(sources, undefined, [])
+      expect(Object.keys(components).sort()).toEqual(['success1', 'success2', 'throwsErrorObject', 'throwsErrorString'])
+      expect(components.success1.error).toBeUndefined()
+      expect(components.success1.value).toBe('foo')
+      expect(components.success2.error).toBeUndefined()
+      expect(components.success2.value).toBe('baq')
+      expect(components.throwsErrorObject.value).toBeUndefined()
+      expect(components.throwsErrorObject.error).toBeInstanceOf(Error)
+      expect(components.throwsErrorObject.error!.message).toBe('bar')
+      expect(components.throwsErrorString.value).toBeUndefined()
+      expect(components.throwsErrorString.error).toEqual({ message: 'baz' })
+    })
+
+    it('measures duration', async () => {
+      const sources = {
+        instant: () => true,
+        delayedResult: () => new Promise((resolve) => setTimeout(resolve, 50)),
+        delayedError: () => new Promise((_resolve, reject) => setTimeout(() => reject('test'), 50)),
+      }
+      const components = await getComponents(sources, undefined, [])
+      expect(components.instant.duration).toBeLessThan(25)
```

Comment:
Why it's 25 here? 
Also maybe it's excessive here but we may mock timeout https://jasmine.github.io/2.0/introduction.html#section-Mocking_the_JavaScript_Timeout_Functions

Reply:
Because lower values lead to unstable tests that fail occasionally. Timer mocking seems to be an elegant solution, I'll consider it, thank you.

Reply:
Unfortunately, `jasmine.clock` isn't suitable:

1. It doesn't work well with asynchronous code: https://github.com/jasmine/jasmine/issues/1659
2. It doesn't adjust JavaScript internal clock (that is used to measure duration in this test)

These issues are solvable, but I don't think that this test is worth spending time on looking for a better `jasmine.clock()` alternative.